### ISSUE TITLE
Inaccurate Results in Cosine Similarity Calculation with Float64

### ISSUE THREAD

**Issue by ShailaShova** [pytorch/pytorch#115786]
### 🐛 Describe the bug

I encountered an unexpected behavior when calculating cosine similarity using PyTorch with float64 data type. Despite using float64 for increased precision, the results significantly deviate from the expected values.

import torch
import numpy as np

# PyTorch computation with float64
a_torch = torch.tensor([2606.66824394, 2477.72226966, 3251.84008903], dtype=torch.float64)
b_torch = torch.tensor([3.46822161e-09, 1.82536693e-09, 8.35245752e-09], dtype=torch.float64)
cos_sim_torch = torch.nn.functional.cosine_similarity(a_torch, b_torch, dim=0)
print(f"PyTorch Cosine Similarity: {cos_sim_torch.item()}") 
That results 0.8399261 

# Equivalent NumPy computation for comparison
a_np = np.array([2606.66824394, 2477.72226966, 3251.84008903])
b_np = np.array([3.46822161e-09, 1.82536693e-09, 8.35245752e-09])
cos_sim_np = np.dot(a_np, b_np) / (np.linalg.norm(a_np) * np.linalg.norm(b_np))
print(f"NumPy Cosine Similarity: {cos_sim_np}")
and the result is 0.91036

I expected the cosine similarity calculated by PyTorch to closely match the result from NumPy, especially since I used float64 for higher precision. I have calculated manually and it gives the result that I get from Numpy calculation. 
I think this discrepancy is concerning, particularly for applications where precision is critical. I would appreciate any insights into whether this is a known limitation or a potential bug, and any guidance on how to address this issue would be highly appreciated. 

### Versions

Collecting environment information...
PyTorch version: 2.1.1
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 12.1 (x86_64)
GCC version: Could not collect
Clang version: 16.0.1
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.13 (main, Aug 25 2022, 18:29:29) [Clang 12.0.0 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1

Versions of relevant libraries:
[pip3] flake8==4.0.1
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.24.2
[pip3] numpydoc==1.4.0
[pip3] onnx==1.12.0
[pip3] torch==2.1.1
[pip3] torchvision==0.14.0
[conda] blas 1.0 mkl
[conda] mkl 2021.4.0 hecd8cb5_637
[conda] mkl-service 2.4.0 py39h9ed2024_0
[conda] mkl_fft 1.3.1 py39h4ab4a9b_0
[conda] mkl_random 1.2.2 py39hb2f4e1b_0
[conda] numpy 1.21.0 pypi_0 pypi
[conda] numpydoc 1.4.0 py39hecd8cb5_0
[conda] torch 2.1.1 pypi_0 pypi
[conda] torchvision 0.14.0 pypi_0 pypi

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki

**Comment by mikaylagawarecki** [comment]
Note that there is an `eps` parameter to `cosine_similarity`, see [docs](https://pytorch.org/docs/stable/generated/torch.nn.functional.cosine_similarity.html)

I modified your script to include the behavior for eps and the results are now the same


```
import torch

a_torch = torch.tensor([2606.66824394, 2477.72226966, 3251.84008903], dtype=torch.float64)
b_torch = torch.tensor([3.46822161e-09, 1.82536693e-09, 8.35245752e-09], dtype=torch.float64)
cos_sim_torch = torch.nn.functional.cosine_similarity(a_torch, b_torch, dim=0)
print(f"PyTorch Cosine Similarity: {cos_sim_torch.item()}")

import numpy as np
print(np.__version__)
a_np = np.array([2606.66824394, 2477.72226966, 3251.84008903])
b_np = np.array([3.46822161e-09, 1.82536693e-09, 8.35245752e-09])
norm_a = np.linalg.norm(a_np)
if norm_a < 1e-8:
    norm_a = 1e-8
norm_b = np.linalg.norm(b_np)
if norm_b < 1e-8:
    norm_b = 1e-8
print(norm_a, norm_b)
cos_sim_np = np.dot(a_np, b_np) / (norm_a * norm_b)
print(f"NumPy Cosine Similarity: {cos_sim_np}")
```

**Comment by mikaylagawarecki** [comment]
@ShailaShova I am closing this since this is the expected behavior, feel free to reopen if you have further questions
